wrangler-transform/src/main/java/io/cdap/wrangler/WranglerErrorUtil.java

```diff
@@ -0,0 +1,202 @@
+      List<Throwable> causalChain = Throwables.getCausalChain(e);
+      for (Throwable t : causalChain) {
+        if (t instanceof ProgramFailureException) {
+          return null;
+        }
+        if (t instanceof DirectiveParseException) {
+          return getProgramFailureExceptionWrappedUnderDirectiveParse(t,
+              errorType != null ? errorType : ErrorType.USER);
+        }
+        if (t instanceof RecipeException) {
+          return getProgramFailureExceptionWrappedUnderRecipeException(t,
+              errorType != null ? errorType : ErrorType.USER);
+        }
+        if (t instanceof PreconditionException) {
+          return getProgramFailureException((PreconditionException) t,
+              errorType != null ? errorType : ErrorType.USER, "Precondition");
+        }
+        if (t instanceof DirectiveExecutionException) {
+          return getProgramFailureException((DirectiveExecutionException) t,
+              errorType != null ? errorType : ErrorType.USER, "Executing-Directive");
+        }
+        if (t instanceof RecipeAlreadyExistsException) {
+          return getProgramFailureException((RecipeAlreadyExistsException) t,
+              errorType != null ? errorType : ErrorType.USER, "RecipeAlreadyExists");
+        }
+        if (t instanceof SchemaNotFoundException) {
+          return getProgramFailureException((SchemaNotFoundException) t,
+              errorType != null ? errorType : ErrorType.USER, "SchemaNotFound");
+        }
+        if (t instanceof WorkspaceNotFoundException) {
+          return getProgramFailureException((WorkspaceNotFoundException) t,
+              errorType != null ? errorType : ErrorType.USER, "WorkspaceNotFound");
+
+        }
+      }
+    }
+    // If no predefined exception found in the causal chain, return generic program failure exception
+    return ErrorUtils.getProgramFailureException(
+        new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN), errorMessage, errorMessage,
+        errorType, false, e);
+  }
+
+  /**
+   * Get a ProgramFailureException with the given error information from {@link Throwable}.
+   * <p>
+   * This method traverses the causal chain of the provided Throwable to find specific exceptions
+   * such as ProgramFailureException, DirectiveLoadException, DirectiveParseException,
+   * DirectiveNotFoundException, and RecordConvertorException. If one of these exceptions is found,
+   * it creates and returns a ProgramFailureException with appropriate error details.
+   * <p>
+   * If no predefined exception is found in the causal chain, a generic ProgramFailureException is
+   * returned.
+   *
+   * @param exception the Throwable to get the error information from
+   * @param errorType the type of error to be used if no specific exception is found
+   * @return a ProgramFailureException with the given error information
+   */
+  private static ProgramFailureException getProgramFailureExceptionWrappedUnderRecipeException(
```

Comment:
There is no need to create these new methods.

You can just check while traversing the original causal chain, if the exception `isTerminal`, if not continue traversing the causal chain and keep on storing/overring the result.

You can create a list of terminal exceptions and error category corresponding to it.

At the end if result is not null return it, otherwise return generic programFailureException.